### PR TITLE
Implement mechanism of override and patch migration files

### LINKED ISSUE #30: Implement amended and fixed files
## Description
According to the issue described here: https://github.com/seqeralabs/nf-tower-cloud/issues/3613 we see a need to implement a mechanism to amend or fix migration files.

Relevant part from the mentioned issue:
The [migration logic](https://github.com/seqeralabs/nf-tower-cloud/blob/master/tower-migrate/src/main/groovy/io/seqera/tower/Migration.groovy) would need to change in order to:
1. Get all files to run and identify those "broken" files with amended and fix companions.
2. Check whether the broken migrations and companions were already applied (requires changes in the [migration tool](https://github.com/seqeralabs/migtool)).
3. Choose the `.fix` or `.amended` files depending on whether the broken file was already applied or not.
4. Send the specific subset of files to execute (requires changes in the migration tool).
5. The migration files execute the given set as usual: the applied files are ignored, the new ones are executed.

### PR DESCRIPTION
**Pull request by weronikasosnowskaseqera**
## Description

Closes #30 

This PR adds functionality of override and patch files described in this issue https://github.com/seqeralabs/nf-tower-cloud/issues/3613.

The migration tool will scan migration files according to the given pattern like before. Besides collecting `migrationEntries`, It will also collect files with `.patch` or `.override` extensions in the file name and put Them in separate lists. In these collections, we will be looking for override or patch files during the `applyMigration` method.

## Scenarios

- WHEN file.sql exists and was executed
  - AND file.override.sql exists -> nothing happens
  - AND file.patch.sql exists -> apply patch
 
- WHEN file.sql exists and was not executed
  - AND file.override.sql exists -> apply override
  - AND file.patch.sql exists -> apply file.sql and apply patch
  - AND file.patch.sql exists AND file.override.sql exists -> apply override
  
- WHEN file.sql doesn't exist but file.patch.sql exists OR file.override.sql exists -> nothing happens
##

### CODE REVIEW

**Comment by weronikasosnowskaseqera** [comment]
I'd like to have more opinions about this behavior mentioned above

> WHEN file.sql exists and was not executed  AND file.fixed.sql exists AND file.amended.sql exists -> apply amended and apply fix

I'm not sure if in this case we would prefer to execute only amended or both amended and fixed file.

**Comment by tcrespog** [comment]
I'm not sure if we should couple the amended/fixed logic to the migtool library. Originally, the proposal described at https://github.com/seqeralabs/nf-tower-cloud/issues/3613 suggested general-purpose changes that would allow to implement the logic on the side of the users of the library.

From the point of view of the tool/library, the idea was to:

1. Add the ability to check if a specific migration file has been applied.
2. Receive a set of migration files to run.

Therefore, with those capabilites in place, from the point of view of users of the tool, the amended/fixed logic could be applied by:

1. Getting all files to run and identifying those "broken" files with amended and fix companions.
2. Checking whether the broken migrations and companions were already applied with migtool.
3. Choose the `.fix` or `.amended` files depending on whether the broken file was already applied or not.
4. Send the specific subset of files to execute.

> I'm not sure if in this case we would prefer to execute only amended or both amended and fixed file.

The idea in this case is to ignore the original (and `.amended`) and apply `.fixed`. In a nutshell:
- Original already applied: run `.amended`.
- Original not applied: run `.fixed`.

**Comment by weronikasosnowskaseqera** [comment]
After talking with @tcrespog decided to close this PR and prepare another one with less changes. We will handle fix and amend files on the platform side.

**Comment by pditommaso** [comment]
> We will handle fix and amend files on the platform side.

Sorry to be late on this, but not sure that a good idea. This tools is used across different projects, tower, licman, groundswell thefore it has more sense to have this logic as a core feature

**Comment by weronikasosnowskaseqera** [comment]
@pditommaso okay, then we can reopen this PR. I tested It with the platform by running migration with all migration files and I added a few fixed, and amended files to test new functionality - all worked well.

When we have this PR merged I will open another PR with more e2e tests on the platform side and actual fixed and amended files described in [this issue](https://github.com/seqeralabs/nf-tower-cloud/issues/3613).

For visibility: I created [similar PR](https://github.com/seqeralabs/nf-tower-cloud/pull/6147) for the platform, logic is the same. If we decided that we will place these changes here, in this repo - I will close that PR.

File: src/main/java/io/seqera/migtool/MigTool.java [hunk 1]
```diff
@@ -321,7 +321,10 @@ else if( locations.startsWith(SOURCE_FILE)) {
         else {
             throw new IllegalArgumentException("Invalid locations prefix: " + locations);
         }
+
         log.debug("Scanned {} migration files, {} amended files and {} fixed files", migrationEntries.size(), amendedEntries.size(), fixedEntries.size());
+        if (amendedEntries.size() != fixedEntries.size())
```
Comment (tcrespog) [inline src/main/java/io/seqera/migtool/MigTool.java:326]: This check might not be enough as there could be a file with only a `.amended` version and another file with only a `.fixed` version that would make it pass. I think it might be better to check that each entry has a matching companion (and report which one doesn't have it).

File: src/test/groovy/io/seqera/migtool/MigToolTest.groovy [hunk 2]
```diff
@@ -220,4 +240,450 @@ class MigToolTest extends Specification {
...
+        tool.existTable(conn, 'AMENDED')
+        !tool.existTable(conn, 'FIXED')
+        tool.existTable(conn, 'YYY')
+        tool.existTable(conn, 'ZZZ')
+
+        cleanup:
+        conn?.close()
+    }
+
+    def 'try to migrate with fix to fixes'() {
```
Comment (tcrespog) [inline src/test/groovy/io/seqera/migtool/MigToolTest.groovy:665]: `with fix having two .fixed suffixes`?

File: src/test/groovy/io/seqera/migtool/MigToolTest.groovy [hunk 3]
```diff
@@ -10,8 +10,7 @@ import spock.lang.Specification
 
 class MigToolTest extends Specification {
 
-
-    def 'should init migtool' () {
+    def setup() {
```
Comment (tcrespog) [inline src/test/groovy/io/seqera/migtool/MigToolTest.groovy:13]: `setup` is called before every test. Why has the method been renamed that way?

File: src/test/groovy/io/seqera/migtool/MigToolTest.groovy [hunk 4]
```diff
@@ -41,6 +40,27 @@ class MigToolTest extends Specification {
         conn?.close()
     }
 
+    def cleanup() {
```
Comment (tcrespog) [inline src/test/groovy/io/seqera/migtool/MigToolTest.groovy:43]: Is a cleanup phase necessary? Being an H2 in-memory database, isn't a fresh one being created on every new connection?

File: src/test/groovy/io/seqera/migtool/MigToolTest.groovy [hunk 5]
```diff
@@ -220,4 +240,450 @@ class MigToolTest extends Specification {
         conn?.close()
     }
 
+    def 'try to apply local file migration with amend and  without fix to one file' () {
```
Comment (tcrespog) [inline src/test/groovy/io/seqera/migtool/MigToolTest.groovy:243]: Might be worth testing the case when there's `.fixed`, but not `.amended` too.

**Review by tcrespog — APPROVED** [review]
I've just noticed a few things that I overlooked initially in the tests, but the rest is fine.

File: src/test/groovy/io/seqera/migtool/MigToolTest.groovy (hunk 4 above)
Comment (weronikasosnowskaseqera) [reply src/test/groovy/io/seqera/migtool/MigToolTest.groovy:43]: not each test, only each run so when I was running all tests at once - It was conflicting. I had to run each test separately. It is a bit ugly, but I didn't find a nicer way at the time.

Comment (tcrespog) [reply src/test/groovy/io/seqera/migtool/MigToolTest.groovy:43]: > not each test, only each run

I'm not sure if I'm understanding correctly, but `cleanup` runs at the end of every test method while `cleanupSpec` runs at the end of all test methods.

> It is a bit ugly, but I didn't find a nicer way at the time

In the test that was already there, I've noticed that the database is not deleted once the connection closes [on purpose](https://github.com/seqeralabs/migtool/blob/6d253dda72fd7bfef74426432e4d1c467feb5f34/src/test/groovy/io/seqera/migtool/MigToolTest.groovy#L18) by using `DB_CLOSE_DELAY=-1`, because the tool might open several connections.

In order to avoid the cleanup phase, we could use the same connection string in the rest of the tests with a different database name each time (`jdbc:h2:mem:test[1,2,3];DB_CLOSE_DELAY=-1`).

**Review by swampie — APPROVED** [review]

**Comment by weronikasosnowskaseqera** [comment]
@pditommaso Can you review this one, please?

**Comment by pditommaso** [comment]
Thanks. I'll check asap

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -30,9 +30,8 @@ jobs:
       - name: Setup Graalvm
         uses: graalvm/setup-graalvm@v1
         with:
-          version: '21.3.0'
-          java-version: '11'
-          components: 'native-image'
+          java-version: '17.0.8'
+          distribution: 'graalvm'
```
Comment (pditommaso) [inline .github/workflows/build.yml:34]: While desiderable, I think Java update should go into a separate PR, unless there's an explicit dependency with this version

File: README.md [hunk 7]
```diff
@@ -10,6 +10,14 @@ In addition, the migration files can be written in Groovy in order to run script
 Only requirement is that migration files follow the pattern `V99__Some_name.[sql|groovy]`, where 
 `99` can be any integer value.
 
+### Fixed and amended files
+
+Beside files matching pattern mentioned above there is also possibility of creating `.fixed.` or `.amended.`, 
+example: `V99__Some_name.[fixed|amended].[sql|groovy]`.
+These can be useful when the original migration file was somehow broken. `.fixed.` file will be applied when original file was already applied,
+and `.amended.` file will be applied on clean instance where original file wasn't yet applied. These two files should be always created together,
+meaning that there can't be `.fixed.` file without `.amended.` file.
```
Comment (pditommaso) [inline README.md:19]: Please provide a better short description how `.fixed.` or `.amended.` should be used. Reading this I would not be able to apply them

**Review by pditommaso — CHANGES_REQUESTED** [review]

File: .github/workflows/build.yml (hunk 6 above)
Comment (swampie) [reply .github/workflows/build.yml:34]: Graalvm for java 11 is throwing an error while building on gh action with the [following error](https://github.com/oracle/graal/issues/3144): I was checking an the only solution I found was to upgrade the java version

Comment (pditommaso) [reply .github/workflows/build.yml:34]: Fair enough

File: README.md [hunk 8]
```diff
@@ -10,6 +10,19 @@ In addition, the migration files can be written in Groovy in order to run script
...
 
+### Fixed and amended files
+
+If you have broken migration scripts, you can use the migration tool to add fixed and amended files to existing migration files. These two files are always created together.
+
+1. If the broken file has already been applied, a fixed file will be used. The fixed file name should follow this pattern:
+   `V99_file_to_fix.fixed.[sql|groovy]`
+
+2. If the broken file hasn't been applied, an amended file will be used instead of the original file. The amended file name should follow this pattern:
+   `V99_file_to_fix.amended.[sql|groovy]`
```
Comment (pditommaso) [inline README.md:21]: There can be more than one pair of `fixed` and `amended` files?

Comment (weronikasosnowskaseqera) [reply README.md:21]: No, I added that to the Warning below.

Comment (pditommaso) [reply README.md:21]: I fear it may be needed, but it can be kept as a future extension. 


Apart this, I think it would be a bit more clear renaming `.amended` -> `.override` (amend sounds like a minor or incremental change) and `.fixed` -> `.patch`

File: README.md [hunk 9]
```diff
@@ -10,6 +10,20 @@ In addition, the migration files can be written in Groovy in order to run script
...
+### Patch and override files
+
+If you have broken migration scripts, you can use the migration tool to add patch and override files to existing migration files. These two files are always created together.
+
+1. If the broken file has already been applied, a patch file will be used. The patch file name should follow this pattern:
+   `V99_file_to_fix.patch.[sql|groovy]`
+
+2. If the broken file hasn't been applied, an override file will be used instead of the original file. The override file name should follow this pattern:
+   `V99_file_to_fix.override.[sql|groovy]`
+
```
Comment (pditommaso) [inline README.md:22]: Better, but I think we need to add a clear description of the behavior because it's essential to avoid ambiguity. This snippet taken from this ticket should be enough

```
WHEN file.sql exists and was executed

AND file.override.sql exists -> nothing happens
AND file.patch.sql exists -> apply patch
WHEN file.sql exists and was not executed

AND file.override.sql exists -> apply override
AND file.patch.sql exists -> apply file.sql and apply patch
AND file.patch.sql exists AND file.override.sql exists -> apply override
WHEN file.sql doesn't exist but file.patch.sql exists OR file.override.sql exists -> nothing happens

```

Comment (pditommaso) [reply README.md:22]: Even tho I'm not understanding the different in the last two statements 

```
AND file.patch.sql exists -> apply file.sql and apply patch
AND file.patch.sql exists AND file.override.sql exists -> apply override
```

Comment (tcrespog) [reply README.md:22]: Both `.patch` and `.override` [must exist](https://github.com/seqeralabs/migtool/blob/c55425d82e7a28b98093214dbe4bd31bd01b93c9/src/main/java/io/seqera/migtool/MigTool.java#L319-L321) together, so I would describe the cases in this way:

```
WHEN file.sql exists and was executed
AND file.patch.sql, file.override.sql pair exists -> apply patch

WHEN file.sql exists and was not executed
AND file.patch.sql, file.override.sql pair exists -> apply override

WHEN file.sql exists
AND only one of file.patch.sql or file.override.sql exists -> illegal state, the migration process is aborted before any migration files are applied
```

Comment (swampie) [reply README.md:22]: so many words for a tool used only by us IMHO

Comment (swampie) [reply README.md:22]:
``` For each migration file that needs to be corrected a pair of .path and .override must exists. The behaviour will follow these rules:

- **WHEN** file.sql exists and was executed **AND** file.patch.sql, file.override.sql pair exists -> apply patch
- **WHEN** file.sql exists and was not executed **AND** file.patch.sql, file.override.sql pair exists -> apply override

If for a given file there is only either a single .patch or a single .override then the tool will throw an illegal state exception stopping the migration process ```

Comment (pditommaso) [reply README.md:22]: ok, let's update the readme

Comment (swampie) [reply README.md:22]: @weronikasosnowskaseqera please...also be sure to change the tower-cloud PR with the files

**Comment by swampie** [comment]
@pditommaso can you please merge and start a release?

**Comment by weronikasosnowskaseqera** [comment]
I want to run It again with tower after the file pattern changes just in case. I will put a comment here when ready.

**Comment by pditommaso** [comment]
> can you please merge and start a release?

we need fix before the current problem, I don't want to overlap things

**Comment by weronikasosnowskaseqera** [comment]
All right, I managed to run It locally with `./gradlew publishToMavenLocal` on the branch with patch and override files https://github.com/seqeralabs/nf-tower-cloud/pull/6160. We need to release this PR to continue with the mentioned PR.

**Review by pditommaso — APPROVED** [review]

**Comment by pditommaso** [comment]
Ok
